New methods [HUGE UPDATE]

kapi2289 opened the pull request:
Added:
- Detecting change of the thread image `onImageChange` #315 
- Fetching `Message` object from message ID `fetchMessageInfo`
- Group managment methods `addGroupAdmins`, `removeGroupAdmins`, `changeGroupApprovalMode`, `acceptUsersToGroup`, `denyUsersFromGroup`,
and detecting them `onAdminsAdded`, `onAdminsRemoved`, `onApprovalModeChange`
- Blocking users `blockUser`, `unblockUser`
- Moving threads `moveThreads` #177 
- Deleting threads `deleteThreads`
- Deleting messages `deleteMessages`
- Muting threads `muteThread`, `muteThreadReactions`, `muteThreadMentions`
and unmuting `unmuteThread`, `unmuteThreadReactions`, `unmuteThreadMentions`
- Marking threads as unread `markAsUnread`
- Marking as read multiple threads `markAsRead`
- Marking thread as spam and deleting it `markAsSpam`
- Waving to a thread `wave`
- Searching for messages `searchForMessageIDs`, `searchForMessages`
- Detecting group call events `onCallStarted`, `onCallEnded`, `onUserJoinedCall`
- Detecting game events `onGamePlayed`
- Few fixes in code
- Polls
  - `Poll`, `PollOption` models
  - Poll methods `createPoll`, `updatePollVote`
  - Fetching poll options `fetchPollOptions`
  - Detecting poll events `onPollCreated`, `onPollVoted`
- Event reminders rebuild
  - `Plan` model
  - Plan methods `createPlan`, `editPlan`, `deletePlan`, `changePlanParticipation`
  - Fetching `Plan` object from event ID `fetchEventInfo`
  - Detecting plan events `onPlanCreated`, `onPlanEnded`, `onPlanEdited`, 
`onPlanParticipation`
- Attachments sending rebuild #194 #308 #321
  - Sending all types of files `sendRemoteFiles`, `sendLocalFiles`
  - Sending from ID `_sendFiles`

kapi2289 commented:
I spent whole day for this update. Hope it will be approved :D

kapi2289 commented:
I think I made too may commits. If you want I will make a new pull request.

madsmtm commented on fbchat/client.py:
```diff
@@ -1519,6 +1650,12 @@ def getThreadIdAndThreadType(msg_metadata):
                                            thread_type=thread_type, ts=ts, metadata=metadata, msg=m)
 
+                    # Thread image change
+                    elif delta.get("class") == "ForcedFetch":
```
I think `ForcedFetch` happens at other times too, an what it means is basically that the client should fetch information about the thread again. So what we should be doing, is fetch the newest events in the thread, and then take the actor_id and image_id from that. Quite expensive though, in terms of performance, but I think it's the only way?

kapi2289 replied:
It's the only event that is returned after changing thread image. 😕 When I looked at requests in browser I found that this information is in graphql event. We could try to fetch that. Second option is to find information by action (message) id but I completely don't know how to fetch that and I think that there isn't any possible way to do that.

kapi2289 commented on fbchat/client.py:
```diff
@@ -1080,13 +1080,82 @@ def removeUserFromGroup(self, user_id, thread_id=None):
+            "thread_fbid": thread_id
+        }
+
+        if type(admin_ids) is not list:
```
I made it like in addUsersToGroup, but we can make helper method and btw make addUserToGroup too.

kapi2289 commented on fbchat/client.py:
```diff
@@ -1080,13 +1080,82 @@ def removeUserFromGroup(self, user_id, thread_id=None):
+
+    def removeGroupAdmins(self, admin_ids, thread_id=None):
+        """
+        Removes group admin from specifed user.
```
Sorry, I'm not perfet at English 😂

kapi2289 commented on fbchat/client.py:
```diff
@@ -1080,13 +1080,82 @@ def removeUserFromGroup(self, user_id, thread_id=None):
+        for i, admin_id in enumerate(admin_ids):
+            data['admin_ids[' + str(i) + ']'] = str(admin_id)
+
+        j = self._post(self.req_url.SAVE_ADMINS, data, fix_request=True, as_json=True)
```
I checked that and there isn't any useful information. But I don't remember if I checked all of this requests.

kapi2289 commented on fbchat/client.py:
```diff
@@ -1080,13 +1080,82 @@ def removeUserFromGroup(self, user_id, thread_id=None):
+
+        j = self._post(self.req_url.SAVE_ADMINS, data, fix_request=True, as_json=True)
+
+    def removeGroupAdmins(self, admin_ids, thread_id=None):
```
Okey, I will make that 😉

kapi2289 commented:
Actually I have problems with logging to my Facebook account by fbchat so I can't make some of this changes. When I'll fix this I'll make changes, thank you!

kapi2289 commented:
I made it! I found request (`_forcedFetch` method) to fetch information about changed thread info. This request can fetch information about every message by `mid` and `thread_id`! I added `onGamePlayed`, changed `moveThread` to `moveThreads` (multiple) and added helper method `_adminStatus`. Thank you!

madsmtm commented on fbchat/graphql.py:
```diff
@@ -127,6 +128,64 @@ def graphql_to_attachment(a):
+    rtn.votes_count = a.get('voters').get('count') if isinstance(a.get('voters'), dict) else a.get('total_count')
+    return rtn
+
+def graphql_to_event(a):
```
Sometimes, `oid` is set as the event ID, see the comment on `client.py/fetchEventInfo`

kapi2289 replied:
Done, I didn't see that 😃

madsmtm commented on fbchat/graphql.py:
```diff
@@ -155,6 +214,7 @@ def graphql_to_user(user):
     if user.get('profile_picture') is None:
         user['profile_picture'] = {}
     c_info = get_customization_info(user)
+    event_reminders = [graphql_to_event(event) for event in user['event_reminders']['nodes']] if user.get('event_reminders') is not None else []
```
The `user.get('event_reminders') is not None` can be replaced by simply `user.get('event_reminders')`. This is true in a few other places, too

kapi2289 replied:
Done

madsmtm commented on fbchat/graphql.py:
```diff
@@ -169,7 +229,8 @@ def graphql_to_user(user):
         name=user.get('name'),
-        message_count=user.get('messages_count')
+        message_count=user.get('messages_count'),
+        event_reminders=event_reminders
```
To avoid longer diffs in the future, it's good practice to always append a trailing comma at the end of these definitions. Just a heads up 😉

kapi2289 replied:
Okey, thanks. Done

madsmtm commented on fbchat/models.py:
```diff
@@ -37,14 +37,17 @@ class Thread(object):
-    def __init__(self, _type, uid, photo=None, name=None, last_message_timestamp=None, message_count=None):
+    #: List of :class:`Event`
+    event_reminders = None
+    def __init__(self, _type, uid, photo=None, name=None, last_message_timestamp=None, message_count=None, event_reminders=[]):
```
See [this discussion](https://stackoverflow.com/q/1132941) on mutable default arguments. In short, `event_reminders` should be set to `None` by default, and initialized to a list when the instance is created

madsmtm replied:
[This](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments) is also a good resource

kapi2289 replied:
Wow, I didn't know that. Thank you!

madsmtm commented on fbchat/models.py:
```diff
@@ -436,6 +439,87 @@ def __repr__(self):
+    title = None
+    #: List of :class:`PollOption`
+    options = None
+    #: Options count
```
Maybe an added note here, stating that all options may not always be available, and that they can be fetched with `Client.fetchPollOptions`

kapi2289 replied:
Done

madsmtm commented on fbchat/models.py:
```diff
@@ -436,6 +439,87 @@ def __repr__(self):
+    def __unicode__(self):
+        return '<PollOption ({}): {} voters={}>'.format(self.uid, repr(self.text), self.voters)
+
+class Event(object):
```
I think we should use the name `Plan` instead of `Event` throughout the code, since that's what Facebook calls them, and since the word "event" could be interpreted as a lot of different things, while "plan" is more specific

kapi2289 replied:
Done

madsmtm commented on fbchat/models.py:
```diff
@@ -436,6 +439,87 @@ def __repr__(self):
+    #: Event location ID
+    location_id = None
+    #: ID of the event creator
+    creator_id = None
```
In a lot of other places, we've used `author_id`, so maybe we should keep it consistent?

kapi2289 replied:
Done

madsmtm commented on fbchat/client.py:
```diff
@@ -1066,7 +1389,7 @@ def addUsersToGroup(self, user_ids, thread_id=None):
         data['log_message_type'] = 'log:subscribe'
 
-        if type(user_ids) is not list:
+        if not isinstance(user_ids, list):
```
As we talked about, it'd be nice to have a helper method for this

kapi2289 replied:
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1316,20 +1733,143 @@ def eventReminder(self, thread_id, time, title, location='', location_id=''):
+
+            j = self._post(self.req_url.CREATE_POLL, data, fix_request=True, as_json=True)
+    
+    def updatePollVote(self, poll_id, option_ids=[], new_options=[]):
```
You type there all selected options, not only added.

kapi2289 replied:
You type there all selected options, not only added.

madsmtm replied:
Oh, okay, that makes sense now that you told me ;)

kapi2289 commented on fbchat/client.py:
```diff
@@ -1386,6 +1940,18 @@ def markAsSeen(self):
         r = self._post(self.req_url.MARK_SEEN, {"seen_timestamp": 0})
         return r.ok
+    
+    def markAsSpam(self, thread_id=None):
```
No, there isn't any way to do that. By the way when I used this option in facebook this only deleted thread. I don't know what this really do.

kapi2289 replied:
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -578,6 +588,53 @@ def searchForThreads(self, name, limit=1):
+            "thread_fbid": thread_id
+        }
+        j = self._post(self.req_url.SEARCH_MESSAGES, data, fix_request=True, as_json=True)
+        snippets = j["payload"]["search_snippets"][query][thread_id]["snippets"]
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -578,6 +588,53 @@ def searchForThreads(self, name, limit=1):
+            "snippetOffset": offset,
+            "snippetLimit": limit,
+            "identifier": "thread_fbid",
+            "thread_fbid": thread_id
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -857,6 +914,58 @@ def fetchImageUrl(self, image_id):
+        message_info = self._forcedFetch(thread_id, mid).get("message")
+        message = graphql_to_message(message_info)
+        if message is None:
+            raise FBChatException('Could not fetch message: {}'.format(mid))
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -857,6 +914,58 @@ def fetchImageUrl(self, image_id):
+            "question_id": poll_id
+        }
+        url_part = urllib.parse.urlencode(data)
+        j = self._post('{}/?{}'.format(self.req_url.GET_POLL_OPTIONS, url_part), fix_request=True, as_json=True)
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -963,25 +1072,187 @@ def sendEmoji(self, emoji=None, size=EmojiSize.SMALL, thread_id=None, thread_typ
+        return self._doSendRequest(data)
 
-        j = self._postFile(self.req_url.UPLOAD, {
+    def _upload(self, file_path, data, mimetype):
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1097,13 +1420,107 @@ def removeUserFromGroup(self, user_id, thread_id=None):
+        thread_id, thread_type = self._getThread(thread_id, None)
+
+        data = {
+            "add": str(admin).lower(),
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1097,13 +1420,107 @@ def removeUserFromGroup(self, user_id, thread_id=None):
+        """
+        self._adminStatus(admin_ids, False, thread_id)
+
+    def changeGroupApprovalMode(self, approval_mode, thread_id=None):
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1316,20 +1733,143 @@ def eventReminder(self, thread_id, time, title, location='', location_id=''):
         url_part = urllib.parse.urlencode(full_data)
 
         j = self._post('{}/?{}'.format(self.req_url.EVENT_REMINDER, url_part), fix_request=True, as_json=True)
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1316,20 +1733,143 @@ def eventReminder(self, thread_id, time, title, location='', location_id=''):
-
-    def setTypingStatus(self, status, thread_id=None, thread_type=None):
+    
+    def editEventReminder(self, event_id, event):
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1281,31 +1698,31 @@ def reactToMessage(self, message_id, reaction):
             "thread_id" : thread_id,
-            "location_id" : location_id,
-            "location_name" : location,
+            "location_id" : event.location_id,
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1316,20 +1733,143 @@ def eventReminder(self, thread_id, time, title, location='', location_id=''):
+
+        if thread_type != ThreadType.GROUP:
+            raise FBchatUserError('Can only create poll in group threads')
+        else:
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1361,7 +1901,17 @@ def markAsDelivered(self, thread_id, message_id):
         return r.ok
 
-    def markAsRead(self, thread_id):
+    def _readStatus(self, read, thread_id):
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1420,6 +1986,182 @@ def removeFriend(self, friend_id=None):
+
+        if location == ThreadLocation.PENDING:
+            location = ThreadLocation.OTHER
+        if location in ThreadLocation:
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1536,6 +2278,21 @@ def getThreadIdAndThreadType(msg_metadata):
 
+                    # Forced fetch
+                    elif delta.get("class") == "ForcedFetch":
+                        mid = delta.get("messageId")
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1545,6 +2302,27 @@ def getThreadIdAndThreadType(msg_metadata):
                                               thread_id=thread_id, thread_type=thread_type, ts=ts, metadata=metadata, msg=m)
 
+                    # Admin added to group thread
+                    elif delta.get("class") == "AdminAddedToGroupThread":
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1545,6 +2302,27 @@ def getThreadIdAndThreadType(msg_metadata):
+                                           thread_type=thread_type, ts=ts, msg=m)
+
+                    # Admin removed from group thread
+                    elif delta.get("class") == "AdminRemovedFromGroupThread":
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -1574,6 +2352,92 @@ def getThreadIdAndThreadType(msg_metadata):
                         self.onMarkedSeen(threads=threads, seen_ts=seen_ts, ts=delivered_ts, metadata=delta, msg=m)
+                    
+                    # Game played
+                    elif delta.get("type") == "instant_game_update":
```
Done

kapi2289 commented on fbchat/client.py:
```diff
@@ -11,6 +11,7 @@
 from .models import *
 from .graphql import *
 import time
+import ast
```
Removed. Now I use json.loads()

madsmtm commented on fbchat/client.py:
```diff
@@ -1779,9 +1773,8 @@ def deletePlan(self, plan):
-        url_part = urllib.parse.urlencode(full_data)
 
-        j = self._post('{}/?{}'.format(self.req_url.PLAN_DELETE, url_part), fix_request=True, as_json=True)
+        j = self._post(self.req_url.PLAN_CHANGE, full_data, fix_request=True, as_json=True)
```
Hmm, seems like you changed `self.req_url.PLAN_DELETE` to `self.req_url.PLAN_CHANGE` by mistake ;)

kapi2289 replied:
No, request for deleting and changing was the same, but I forgot about it and when changing Event to Plan I wrote PLAN_DELETE, hah.

madsmtm commented on fbchat/client.py:
```diff
@@ -2217,6 +2198,7 @@ def _parseMessage(self, content):
 
         for m in content["ms"]:
             mtype = m.get("type")
+            print(json.dumps(m, indent=4))
```
You left some debug code there ;)

kapi2289 replied:
Oh, sorry 😂 can this stay commented in origin code?

madsmtm commented on fbchat/graphql.py:
```diff
@@ -214,7 +214,7 @@ def graphql_to_user(user):
         user['profile_picture'] = {}
     c_info = get_customization_info(user)
-    plan = graphql_to_plan(user['event_reminders']['nodes'][0]) if user.get('event_reminders') else None
+    plan = graphql_to_plan(user['event_reminders']['nodes'][0]) if user.get('event_reminders', dict()).get('nodes') else None
```
Be aware that `user.get('event_reminders', dict()).get('nodes')` can still raise an `AttributeError`, on data that looks like e.g:
```py
user = {
    'event_reminders': None
}
```
But it might not be necessary to check this case

kapi2289 replied:
I've never seen `None` in `event_reminders`, but maybe I'm wrong.

madsmtm commented:
I couldn't explain what I meant, so I went ahead and changed `sendLocalFiles` and `sendRemoteFiles`, so you don't need to input the filetypes.

madsmtm commented:
I've made a bunch of tests for the methods you've added, though the `search` methods are still missing tests (haven't found a good way to set up tests for that yet). But I've tested them myself, and they look good. I'll merge this now, then and in a few days I'll make a new release
